src/print/index.ts

```diff
@@ -564,33 +532,113 @@ function printChildren(path: FastPath, print: PrintFn, surroundingLines = true):
      * desired to have text directly wrapping a mustache tag without additional whitespace.
      */
     function flush() {
-        if (!isEmptyGroup(currentGroup)) {
-            trimLeft(currentGroup);
-            trimRight(currentGroup);
-            childDocs.push(fill(currentGroup));
+        let groupDocs = currentGroup.map((item) => item.doc);
+        const groupNodes = currentGroup.map((item) => item.node);
+
+        for (let doc of extractOutermostNewlines(groupDocs)) {
+            outputChildDoc(doc, groupNodes)
         }
+
         currentGroup = [];
     }
 
-    path.each(childPath => {
+    path.each((childPath) => {
         const childNode = childPath.getValue() as Node;
         const childDoc = childPath.call(print);
 
         if (isInlineNode(childNode)) {
-            currentGroup.push(childDoc);
+            currentGroup.push({ doc: childDoc, node: childNode });
         } else {
             flush();
-            childDocs.push(concat([breakParent, childDoc]));
+
+            outputChildDoc(isLine(childDoc) ? childDoc : concat([breakParent, childDoc]), [
+                childNode,
+            ]);
         }
     }, 'children');
 
     flush();
+    lastChildDocProduced();
 
-    return concat([
-        surroundingLines ? softline : '',
-        join(hardline, childDocs),
-        surroundingLines ? dedent(softline) : '',
-    ]);
+    return childDocs
+}
+
+/**
+ * Print the nodes in `path` indented and with leading and trailing newlines.
+ */
+function printIndentedWithNewlines(
+    path: FastPath,
+    print: PrintFn
+): Doc {
+    return indent(
+        concat([softline, ...trim(printChildren(path, print), isLine), dedent(softline)]),
+    );
+}
+
+/**
+ * Print the nodes in `path` indented but without adding any leading or trailing newlines.
+ */
+function printIndentedPreservingWhitespace(
+    path: FastPath,
+    print: PrintFn
+) {
+    return indent(concat(dedentFinalNewline(printChildren(path, print))))
+}
+
+/**
+ * Split the text into words separated by whitespace. Replace the whitespaces by lines,
+ * collapsing multiple whitespaces into a single line. 
+ * 
+ * If the text starts or ends with multiple newlines, those newlines should be "keepIfLonely" 
+ * since we want double newlines in the output.
+ */
+function splitTextToDocs(text: string): Doc[] {
+    let docs: Doc[] = text.split(/[\t\n\f\r ]+/)
```

Comment:
I haven't investigated this thoroughly, but [prettier is using](https://github.com/prettier/prettier/blob/master/src/language-html/utils.js#L17) this regexp to look for whitespace so I thought it best to copy that. 
It seems like \s includes [a bunch of other unicode characters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes). It may or may not be correct to include these too.

Reply:
If prettier does it like this, we do it like this, too. Thanks for the info!